File: python/arena.py
from typing import Callable, Generic, TypeVar

A = TypeVar('A')
B = TypeVar('B')
C = TypeVar('C')

class Arena(Generic[A, B]):
    def __init__(self, pos: A, dis: Callable[[A], B]):
        self.pos = pos
        self.dis = dis

class Lens(Generic[Dom, Cod]):
    def __init__(self, observe: Callable[[Dom], Cod], interpret: Callable[[Dom, Callable[[Cod], Cod]], Callable[[Dom], Dom]]):
        self.observe = observe
        self.interpret = interpret

def id_lens(a: Arena[A, A]) -> Lens[A, A]:
    return Lens(lambda x: x, lambda p, f: lambda x: f(x))

def compose(lens23: Lens[A2, A3], lens12: Lens[A1, A2]) -> Lens[A1, A3]:
    obs = lambda p: lens23.observe(lens12.observe(p))
    int_ = lambda p, f: lens12.interpret(p, lambda x: lens23.interpret(x, lambda y: f(y)))
    return Lens(obs, int_)

Display = tuple[A, Callable[[A], B]]
AsFunctor = tuple[A, Callable[[Callable[[A], B]], C]]

A1 = TypeVar('A1', bound=Arena)
A2 = TypeVar('A2', bound=Arena)
B1 = TypeVar('B1', bound=Arena)
B2 = TypeVar('B2', bound=Arena)

Lens = Tuple[Callable[[Tuple[Tuple[A1, A2], Tuple[B1, B2]]], Tuple[Tuple[A1, B1], Callable[[Dis], Tuple[A2, B2]]]], Callable[[Tuple[Tuple[A1, A2], Tuple[B1, B2]], Dis], Dis]]

def duoidal(a1: A1, a2: A2, b1: B1, b2: B2) -> Lens:
    x = (combine(a1, a2), combine(b1, b2))
    y = (pair(a1, b1), combine(pair(a2, b2)))

    def o(p: Tuple[Tuple[A1, A2], Tuple[B1, B2]]) -> Tuple[Tuple[A1, B1], Callable[[Dis], Tuple[A2, B2]]]:
        p1, p2 = p
        q1, q2 = p1

        def pp(d: Dis) -> Tuple[A2, B2]:
            return (apply(p2, fst(d)), apply(q2, snd(d)))

        return (p1, q1), pp

    def i(p: Tuple[Tuple[A1, A2], Tuple[B1, B2]], d: Dis) -> Dis:
        p1, p2 = p
        q1, q2 = p1
        de1, de2 = d
        return (pair(fst(de1), fst(de2)), pair(snd(de1), snd(de2)))

    return o, i

File: ocaml/lib/arena.ml
type 'a arena = {
  pos : 'a;
  dis : 'a -> 'b;
}

type ('dom, 'cod) lens = {
  observe : 'dom -> 'cod;
  interpret : 'dom -> ('cod -> 'cod) -> ('dom -> 'dom);
}

let id_lens (a : 'a arena) : ('a arena, 'a arena) lens =
  {observe = (fun x -> x); interpret = (fun _ f x -> f x)}

let ( <.> ) (lens23 : ('a2 arena, 'a3 arena) lens) (lens12 : ('a1 arena, 'a2 arena) lens) : ('a1 arena, 'a3 arena) lens =
  let obs p = lens23.observe (lens12.observe p) in
  let int p f = lens12.interpret p (fun x -> lens23.interpret (lens12.observe p) (f (lens23.observe x))) in
  {observe = obs; interpret = int}

type 'a display = 'a * ('a -> 'b)
type 'a as_functor = 'a * (('a -> 'b) -> 'c)

module Lens = struct
  type ('a, 'b) t = Lens of ('a -> 'b * ('b -> 'a))

  let duoidal a1 a2 b1 b2 =
    let x = (Arena.op_@@_ a1 a2, Arena.op_*_ (Arena.op_@@_ b1 b2)) in
    let y = (Arena.op_*_ (a1, b1), Arena.op_@@_ (Arena.op_*_ (a2, b2))) in
    let o ((p1, p2), (q1, q2)) =
      let pp = ((p1, q1), fun d -> (Arena.get_fst d p2, Arena.get_snd d q2)) in
      pp
    in
    let i ((p1, p2), (q1, q2)) ((de1, de2) : Arena.dis y (o ((p1, p2), (q1, q2)))) =
      ((Arena.op_**_ (Arena.get_fst de1, Arena.get_fst de2)),
       (Arena.op_**_ (Arena.get_snd de1, Arena.get_snd de2)))
    in
    Lens (o, i)
end